src/include/device.h

```
@@ -12,9 +12,13 @@
 #include "nccl.h"
 #include "rccl_float8.h"
 #if ROCM_VERSION >= 60000
-   // hip_bf16.h should be used from ROCm 6.0
+  // hip_bf16.h should be used from ROCm 6.0
   #include <hip/hip_bf16.h>
-  typedef __hip_bfloat16 hip_bfloat16;
+  #ifndef _HIP_INCLUDE_HIP_AMD_DETAIL_HIP_BFLOAT16_H_
+    #define _HIP_INCLUDE_HIP_AMD_DETAIL_HIP_BFLOAT16_H_
+    #define _HIP_BFLOAT16_H_
+    typedef __hip_bfloat16 hip_bfloat16;
+  #endif
```

I'm concerned that this causes `hip_bfloat16` to resolve to a different type in different compilation units depending on what other headers they include, and also forces a particular inclusion order. I understand that it's a workaround, but are there any concerns about possible consequences here? I admit that it might just be fine, practically speaking.

---

Yeah. This is not perfect. The ultimate way is to fix all rocm files that uses the old hip_bfloat16.h. You concern is legit. If some other rocm unit is processed first then the "typedef __hip_bfloat16 hip_bfloat16" is muted and rccl will use the old definition of blfoat16

---

I think we can error out in that case, avoiding the cases where rccl is silently using old bfloat16 and impacts  performance.